### PR TITLE
200 - Implement IdsMessage

### LINKED ISSUE #200: IdsMessage: Add new web component
Add the message component https://main-enterprise.demo.design.infor.com/components/message/list

**Phase 1 Features (MVP)**
- Types (error / alert ect)
- Allowed tags/Xss
- Check issue tracker 
- Veto/Buttons ect
- Check https://main-enterprise.demo.design.infor.com/components/message

**Future Features (Not Now)**
NA

**Suggested Internal Markup**
```js
<ids-message>
  <div slot="title"></div>
  <div slot="message"></div>
  <div slot="buttons"></div>
</ids-message>
```

**Suggested Dependencies**
- ids-modal
- ids-text
- ids-button
- ids-alerts

**Settings**
- type
- allowed tags

**Design**
- https://www.figma.com/file/ri2Knf3KchdfdzRAeds0Ab/IDS-Mobility-v4.6?node-id=1%3A5740

**Events**
- beforeopen/open/afteropen
- beforeclose/close/afterclose

**Other**
Pass the following checklist ordered in priority (some may be left for later) https://github.com/infor-design/enterprise-wc/blob/main/doc/CHECKLIST.md

### PR DESCRIPTION
**Pull request by EdwardCoyle**
**Explain the details for making this change. What existing problem does the pull request solve?**
This PR adds the following:
- Adds the IdsMessage Component (comparable to 4.x Message Component)
- Adds the IdsModalButton Component
- Ports part the 4.x `xssUtils` library into this repo, renaming it to `IdsXssUtils`, along with its test suite
- Adds an `IdsXssMixin`, which mixes in a property for storing ignored tags in HTML strings, and a method that can remove unwanted HTML tags from a string based on a components stored ignored tags.

**Related github/jira issue (required)**:
Closes #200 

**Steps necessary to review your pull request (required)**:
- Pull, build, run.
- Open http://localhost:4300/ids-message.  Ensure Message can open/close, that the Modal Buttons work to close.
- Open http://localhost:4300/ids-message/types.  Try out all available message type examples, allowed/disallowed HTML tags examples.
- Open http://localhost:4300/ids-message/standalone-css.  Standalone CSS styles should be working
- Try http://localhost:4300/ids-modal and ensure the Modal looks OK with the new Modal Button at the bottom.
- All tests should pass

**Included in this Pull Request**:
- [x] An e2e or functional test for the bug or feature.
- [x] A note to the change log.

### CODE REVIEW

File: doc/CHANGELOG.md [hunk 1]
```diff
@@ -55,6 +55,10 @@
   - Now exists as its own component
   - Markup is `<ids-menu-button></ids-menu-button>`
   - Implicitly linked to a menu by using `menu="my-menu"` CSS selector attribute, or direct element reference in JS
+- `[Message]` The Message is now a web component called ids-menu
```
Comment (tmcconechy) [inline doc/CHANGELOG.md:58]:
```suggestion
- `[Message]` The Message is now a web component called ids-message
```

File: src/ids-message/ids-message.js [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+} from '../ids-base/ids-element';
+
+import IdsModal from '../ids-modal';
+import IdsIcon from '../ids-icon';
+
+import { attributes } from '../ids-base/ids-attributes';
+import { IdsStringUtils } from '../ids-base/ids-string-utils';
+import IdsDOMUtils from '../ids-base/ids-dom-utils';
+
+// @ts-ignore
```
Comment (tmcconechy) [inline src/ids-message/ids-message.js:14]: Dont think we need this anymore

File: src/ids-message/ids-message.js [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+import styles from './ids-message.scss';
+
+// Types of status that can be applied to message components
+const MESSAGE_STATUSES = [
+  'none', 'default', 'error', 'alert', 'success', 'info'
+];
+
+// Attributes that apply to message components
+const MESSAGE_ATTRIBUTES = [
```
Comment (tmcconechy) [inline src/ids-message/ids-message.js:23]: From the other PR think we decided to move this down.

**Review by tmcconechy — CHANGES_REQUESTED** [review]
Didnt want to request changes as this is very nice work but a couple small things we could improve.

1. see comments for nits
2. on http://localhost:4300/ids-message/types the "large message" doesnt do anyting
3. From an API standpoint i dont quite understand how it opens?

```
const messageEl = document.querySelector('#my-message');
messageEl.messageTitle = 'Different Title';
messageEl.message = 'Alternate Message';
messageEl.status = 'success';
```

Was expecting maybe a show() function or similar

3. A bit of a funny animation on http://localhost:4300/ids-message/types in two places. A) When you open the page the overlay seems to "blink". And when you click cancel the press animation is a bit too dark? Should it just be azure too there?

https://drive.google.com/file/d/1umRr4IoncLAFerlDZTFHkZV3MPTQnj4Z/view

4. The borders are not the same color as https://main-enterprise.demo.design.infor.com/components/message/example-index.html (between the buttons) they look a little dark

**Comment by EdwardCoyle** [comment]
@tmcconechy thanks, I'll address all of those points shortly.  I wanted to answer the question about the API:  this component extends Modal, so you can call `show()/hide()` the same way -- we didn't have to re-implement that at the message level this time around.

In the example code, all the messages are being "opened" by way of [setting the `target` property](https://github.com/infor-design/enterprise-wc/blob/ids-message/app/ids-message/example.js#L10), the same way the Modal example does.  There's also [a line in the sample](https://github.com/infor-design/enterprise-wc/blob/ids-message/app/ids-message/example.js#L16) that will use `hide()` when the Modal buttons are clicked.

**Comment by tmcconechy** [comment]
OK, Then maybe we should have a `show` method because they may want to show a message on a system error or something where they cannot link the message to a button.

**Comment by EdwardCoyle** [comment]
There IS a `show()` method already.  It's not in the `IdsMessage` source code because it's already on the API due to the Message extending the Modal

**Comment by tmcconechy** [comment]
Yeah maybe just show an example in the README.md what it would look like with .show()

**Review by rob2d — COMMENTED** [review]
@EdwardCoyle overall things are looking really good in terms of conventions etc. I think I discovered a few issues though. 

I'm not sure whether we should log an issue or we write TODOs for some of these but: 

1. If you press tab throughout the app the user can still tab to buttons beyond the popover. This could cause some pretty nasty UX states if left uncontrolled in client apps if they aren't noticing that is possible.
1. The "No Buttons" option in `types.html` does not have a click off; I see there's an issue here already, wasn't sure if there was something UI wise that I should know to click off but seems on the example it's a bit unideal to not be able to exit at all without reload.
1. The "Large Message" button in `types.html` does not yet have functionality -- was not sure whether that was meant to do later but didn't see a `TODO` file here.
1. this is a bit of a nit but it feels like the close button closes itself without any visual feedback (similar to other buttons). Maybe the popover close anim is too fast, or maybe we can solve with a simple anim delay if this is a hassle to add retrospectively?

Outside of that code-wise not seeing any issues so the rest of stuff is looking pretty good!

**Comment by EdwardCoyle** [comment]
@rob2d Thanks for the feedback!  I'm addressing pretty much all those points on this PR except for this one:

> If you press tab throughout the app the user can still tab to buttons beyond the popover. This could cause some pretty nasty UX states if left uncontrolled in client apps if they aren't noticing that is possible.

I'm aware of this one and our 4.x Modal(s) did handle this situation.  I've raised #267 for this because I think solution is not necessarily limited to the Modal/Message.

File: src/ids-modal/ids-modal.js [hunk 4]
```diff
@@ -220,17 +363,17 @@ class IdsModal extends mix(IdsElement).with(...appliedMixins) {
     }
 
     this.visible = false;
-    this.#removeOpenEvents();
+    this.removeOpenEvents();
   }
 
-  #applyOpenEvents() {
+  applyOpenEvents() {
```
[the remark this answers is not in the document]
Comment (EdwardCoyle) [reply src/ids-modal/ids-modal.js:369]: I'm going to be fixing this one by implementing the `IdsPopupOpenEventsMixin` here.

File: src/ids-modal/ids-modal.js [hunk 5]
```diff
@@ -220,17 +363,17 @@ class IdsModal extends mix(IdsElement).with(...appliedMixins) {
...
     /* istanbul ignore next */
-    this.listen('Escape', this.container, () => {
+    this.listen('Escape', this, () => {
       this.hide();
     });
   }
 
-  #removeOpenEvents() {
+  removeOpenEvents() {
```
[the remark this answers is not in the document]
Comment (EdwardCoyle) [reply src/ids-modal/ids-modal.js:376]: Also fixing this one by implementing `IdsPopupOpenEventsMixin`

File: src/ids-popup/ids-popup.js [hunk 6]
```diff
@@ -190,10 +190,14 @@ class IdsPopup extends mix(IdsElement).with(
    * @returns {void}
    */
   handleEvents() {
+    /*
```
[the remark this answers is not in the document]
Comment (EdwardCoyle) [reply src/ids-popup/ids-popup.js:193]: Yeah just accidentally left this one in

**Comment by EdwardCoyle** [comment]
@tmcconechy @rob2d I addressed pretty much everything except what's been punted to new issues:
- Added some docs for using `show()/hide()` from the Modal API
- Fixed the unwanted "blink" animation on the first Modal/Message open
- Added a slight delay to show the ripple effect when clicking Modal buttons.  I brightened up the secondary-button's ripple effect so it's not so dark, but kept it the secondary color instead of Azure.  Think that's ok?
- Made it possible to click out of the modal (also added a TODO for supporting the `keep-open` attribute like PopupMenu does)
- Did some code re-arranging among Modal, PopupMenu, and the new Popup mixins to share a bit more nicely
- Added/moved tests where it made sense

Take another look if you like 👍

**Review by tmcconechy — APPROVED** [review]
Looks great!

**Review by rob2d — APPROVED** [review]
Good stuff Ed!  👍

**Review by tmcconechy — DISMISSED** [review]

**Review by rob2d — APPROVED** [review]

**Comment by vivz753** [comment]
http://localhost:4300/ids-message/types
 
 For the Confirmation Example, should `Yes` button ripple with the primary (blue) color? I figure since `Yes` is the first highlighted action that it's meant to be the primary action, thus have the primary colors, which I noticed to be blue for the other examples, however it is grey in this example.

**Comment by EdwardCoyle** [comment]
@vivz753 Good catch!  That one's ok -- in that particular example "Yes" is a secondary button, so its ripple color matches the gray text and border

**Review by tmcconechy — APPROVED** [review]

**Pull request merged**